lib/core/src/Cardano/Wallet/Primitive/Model.hs

```diff
@@ -490,99 +508,118 @@ isOurTx tx u
+    
+isOursState :: IsOurs s entity => entity -> State s Bool
+isOursState x = isJust <$> state (isOurs x)
 
 {-------------------------------------------------------------------------------
-                               Internals
+    Modification of UTxO
 -------------------------------------------------------------------------------}
 
--- | Prefiltering returns all transactions of interest for the wallet. A
--- transaction is a matter of interest for the wallet if:
---
---    - It has known input(s)
---    - and/or It has known output(s)
---
--- In practice, most transactions that are of interest have an output to the
--- wallet but some may actually have no change output whatsoever and be only
+-- | Apply a 'Block' to the 'UTxO'.
+--
+-- Here, we assume that address discovery has already been performed and
+-- that the address discovery state @s@ identifies all our output addresses
+-- in the given 'Block'.
+--
+-- A 'FilteredBlock' is returned in addition to the new 'UTxO' set.
+-- This 'FilteredBlock' includes those transactions and delegations
+-- that are in the given 'Block' and that are also relevant to the wallet,
+-- i.e. they have
+-- 
+-- * Outputs with known addresses
+-- * Inputs referrring to known outputs.
+-- 
+-- In practice, most transactions that are relevant have a change output to the
+-- wallet, but some may actually have no change output whatsoever and be only
 -- linked to the wallet by their inputs.
 --
--- In order to identify transactions that are ours, we do therefore look for
--- known inputs and known outputs. However, we can't naively look at the domain
--- of the utxo constructed from all outputs that are ours (as the specification
--- would suggest) because some transactions may use outputs of a previous
--- transaction within the same block as an input. Therefore, looking solely at
--- the final 'dom (UTxO ⊳ oursOuts)', we would be missing all intermediate txs
--- that happen from _within_ the block itself.
---
--- As a consequence, we do have to traverse the block, and look at transactions
--- in order, starting from the known inputs that can be spent (from the previous
--- UTxO) and collect resolved tx outputs that are ours as we apply transactions.
+-- As inputs are given as references to outputs (no address, no coin value),
+-- we have to traverse all transactions in the block in order to
+-- discover the outputs that belong to us and be able to infer that the
+-- corresponding inputs belong to us as well.
 applyBlockToUTxO
     :: (IsOurs s Address, IsOurs s RewardAccount)
     => Block
-    -> UTxO
     -> s
-    -> ((FilteredBlock, UTxO), s)
-applyBlockToUTxO b u0 = runState $ do
-    delegations <- mapMaybeM ourDelegation (b ^. #delegations)
-    (transactions, ourU) <- foldM applyOurTx (mempty, u0) (b ^. #transactions)
-    return (FilteredBlock {delegations, transactions}, ourU)
+    -> UTxO
+    -> (FilteredBlock, UTxO)
+applyBlockToUTxO block s u0 =
```

Comment:
don't we want to use bangs here?

Reply:
Unfortunately, the shotgun approach to using bangs usually does not improve performance. (The best place to put bangs is on record fields of data types anyway.)

In this case, it's the responsibility of `applyBlockToUTxO` to make sure that the accumulator is evaluated to WHNF; this is indeed the case.

That said, I soon intend to take another look at whether any of our long-lasting state is in normal form.